File: app/src/main/java/com/bongorian/signa1/FrameHistory.java
package com.bongorian.signa1;

import java.util.*;
import java.util.function.Supplier;

/** Bounded handoff from GL submission to UI acknowledgement and shutter reservation.
 * Payloads are writable only while acquired by GL. A shutter pins the acknowledged payload
 * synchronously BEFORE posting work to GL, so queued camera callbacks cannot overwrite it. */
final class FrameHistory<T> {
    static final class Slot<T> {final T value;private long timestamp;private int readers;private boolean writing;Slot(T value){this.value=value;}}
    static final class Lease<T> {final T value;final long timestamp;private final Slot<T> slot;private final long epoch;private boolean released;Lease(Slot<T> slot,long epoch){this.slot=slot;value=slot.value;timestamp=slot.timestamp;this.epoch=epoch;}}
    private final List<Slot<T>> slots=new ArrayList<>();private long acknowledged,epoch;
    FrameHistory(int capacity,Supplier<T> factory){if(capacity<2)throw new IllegalArgumentException("Frame capacity");for(int i=0;i<capacity;i++)slots.add(new Slot<>(factory.get()));}
    synchronized Slot<T> acquire(){for(Slot<T> slot:slots)if(!slot.writing&&slot.readers==0&&(slot.timestamp==0||slot.timestamp<acknowledged)){slot.writing=true;slot.timestamp=0;return slot;}return null;}
    synchronized void publish(Slot<T> slot,long timestamp){if(!slot.writing||timestamp<=0)throw new IllegalStateException("Frame publication");slot.timestamp=timestamp;slot.writing=false;}
    synchronized void abandon(Slot<T> slot){slot.writing=false;slot.timestamp=0;}
    synchronized boolean acknowledge(long timestamp){if(timestamp<=0)return false;for(Slot<T> slot:slots)if(!slot.writing&&slot.timestamp==timestamp&&timestamp>=acknowledged){acknowledged=timestamp;return true;}return false;}
    synchronized Lease<T> reserve(){return reserve(acknowledged);}
    synchronized Lease<T> reserve(long timestamp){if(timestamp<=0)return null;for(Slot<T> slot:slots)if(!slot.writing&&slot.timestamp==timestamp){slot.readers++;return new Lease<>(slot,epoch);}return null;}
    synchronized boolean valid(Lease<T> lease){return lease!=null&&!lease.released&&lease.epoch==epoch&&lease.slot.timestamp==lease.timestamp;}
    synchronized void release(Lease<T> lease){if(lease!=null&&!lease.released){lease.released=true;lease.slot.readers--;}}
    synchronized long acknowledged(){return acknowledged;}
    synchronized void clear(){epoch++;acknowledged=0;for(Slot<T> slot:slots){slot.timestamp=0;slot.writing=false;}}
    List<T> values(){List<T> values=new ArrayList<>();for(Slot<T> slot:slots)values.add(slot.value);return Collections.unmodifiableList(values);}
}

File: app/src/main/kotlin/com/bongorian/signa1/FrameHistory.kt
package com.bongorian.signa1

import java.util.Collections
import java.util.function.Supplier

/**
 * Bounded handoff from GL submission to UI acknowledgement and shutter reservation. Payloads are
 * writable only while acquired by GL. A shutter pins the acknowledged payload synchronously BEFORE
 * posting work to GL, so queued camera callbacks cannot overwrite it.
 */
internal class FrameHistory<T : Any>(capacity: Int, factory: Supplier<T>) {
    internal class Slot<T : Any>(val value: T) {
        internal var timestamp: Long = 0
        internal var readers = 0
        internal var writing = false
    }

    internal class Lease<T : Any>(internal val slot: Slot<T>, internal val epoch: Long) {
        val value: T
        val timestamp: Long
        internal var released = false

        init {
            value = slot.value
            timestamp = slot.timestamp
        }
    }

    internal val slots: MutableList<Slot<T>> = ArrayList<Slot<T>>()
    private var acknowledged: Long = 0
    private var epoch: Long = 0

    init {
        require(capacity >= 2) { "Frame capacity" }
        for (i in 0..<capacity) slots.add(Slot<T>(factory.get()))
    }

    @Synchronized
    fun acquire(): Slot<T>? {
        for (slot in slots) if (
            !slot.writing &&
                slot.readers == 0 &&
                (slot.timestamp == 0L || slot.timestamp < acknowledged)
        ) {
            slot.writing = true
            slot.timestamp = 0
            return slot
        }
        return null
    }

    @Synchronized
    fun publish(slot: Slot<T>, timestamp: Long) {
        check(!(!slot.writing || timestamp <= 0)) { "Frame publication" }
        slot.timestamp = timestamp
        slot.writing = false
    }

    @Synchronized
    fun abandon(slot: Slot<T>) {
        slot.writing = false
        slot.timestamp = 0
    }

    @Synchronized
    fun acknowledge(timestamp: Long): Boolean {
        if (timestamp <= 0) return false
        for (slot in slots) if (
            !slot.writing && slot.timestamp == timestamp && timestamp >= acknowledged
        ) {
            acknowledged = timestamp
            return true
        }
        return false
    }

    @Synchronized
    fun reserve(): Lease<T>? {
        return reserve(acknowledged)
    }

    @Synchronized
    fun reserve(timestamp: Long): Lease<T>? {
        if (timestamp <= 0) return null
        for (slot in slots) if (!slot.writing && slot.timestamp == timestamp) {
            slot.readers++
            return Lease<T>(slot, epoch)
        }
        return null
    }

    @Synchronized
    fun valid(lease: Lease<T>?): Boolean {
        return lease != null &&
            !lease.released &&
            lease.epoch == epoch &&
            lease.slot.timestamp == lease.timestamp
    }

    @Synchronized
    fun release(lease: Lease<T>?) {
        if (lease != null && !lease.released) {
            lease.released = true
            lease.slot.readers--
        }
    }

    @Synchronized
    fun acknowledged(): Long {
        return acknowledged
    }

    @Synchronized
    fun clear() {
        epoch++
        acknowledged = 0
        for (slot in slots) {
            slot.timestamp = 0
            slot.writing = false
        }
    }

    fun values(): MutableList<T> {
        val values: MutableList<T> = ArrayList<T>()
        for (slot in slots) values.add(slot.value)
        return Collections.unmodifiableList<T>(values)
    }
}
